modules/internal/ChapelDebugPrint.chpl

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2004-2016 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* Debug printing for internal modules.
+
+   In order to work around problems with resolution order
+   (commonly, stdout not defined in methods in DefaultRectangular),
+   this module implements a debug printing facility that can
+   be used before IO.chpl is resolved (or initialized).
+ */
+
+module ChapelDebugPrint {
+
+  // This routine is called in DefaultRectangular in order
+  // to report an out of bounds access for a halt. A normal
+  // call to halt with a tuple can't be made because of module
+  // order issues.
+  pragma "no doc"
+  proc _stringify_index(tup:?t) where isTuple(t)
+  {
+    var str = "(";
+
+    for param i in 1..tup.size {
+      if i != 1 then str += ", ";
+      str += tup[i]:string;
+    }
+
+   str += ")";
+
+    return str;
+
+  }
+
+  proc chpl_debug_stringify(args...) : string {
+    var str = "";
+    for param i in 1..args.size {
+      var tmp = args(i);
+      if isNumericType(tmp.type) || isBoolType(tmp.type) {
+        str += tmp:string;
+      } else if isRangeType(tmp.type) {
+        str += tmp.lo:string;
+        str += "..";
+        str += tmp.hi:string;
+      } else if isTupleType(tmp.type) {
+        str += _stringify_index(tmp);
```

Comment:
You could also cast the tuple to string directly, except it would be a resolution error if any tuple elements don't have a cast to string defined, instead of "?".